proxy.go

```diff
@@ -1,33 +1,80 @@
 package gondola
 
 import (
+	"context"
+	"fmt"
 	"io"
 	"log"
 	"net/http"
 	"net/http/httputil"
 	"net/url"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
 )
 
-func New(cfgReader io.Reader) {
+func Load(reader io.Reader) (*Config, error) {
 	cfg := &Config{}
-	if _, err := cfg.Load(cfgReader); err != nil {
-		log.Fatal("Error loading config file")
+	c, err := cfg.Load(reader)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func NewServer(cfg *Config) (*http.Server, error) {
+	s := &http.Server{
+		Addr:              ":" + cfg.Proxy.Port,
+		ReadHeaderTimeout: time.Duration(cfg.Proxy.ReadHeaderTimeout) * time.Second,
 	}
 
 	for _, b := range cfg.Upstreams {
-		upstream, err := url.Parse(b.Address)
+		pp, err := url.Parse(b.Target)
 		if err != nil {
-			log.Fatal("Error parsing upstream address")
+			return nil, fmt.Errorf("error parsing upstream address: %w", err)
 		}
-		proxy := httputil.NewSingleHostReverseProxy(upstream)
-		http.Handle("/", proxy)
-		// TODO: implement virtual hosts
-		break
+
+		proxy := httputil.NewSingleHostReverseProxy(pp)
+		http.HandleFunc(b.HostName+"/", func(w http.ResponseWriter, r *http.Request) {
+			proxy.ServeHTTP(w, r)
+		})
```

Comment:
> Patterns may optionally begin with a host name, restricting matches to URLs on that host only. Host-specific patterns take precedence over general patterns, so that a handler might register for the two patterns "/codesearch" and "codesearch.google.com/" without also taking over requests for "http://www.google.com/".

cf. https://pkg.go.dev/net/http#ServeMux